parsl/providers/local/local.py

```diff
@@ -158,26 +209,47 @@ def submit(self, command, tasks_per_node, job_name="parsl.localprovider"):
 
         job_id = None
         remote_pid = None
-        if (self.move_files is None and not isinstance(self.channel, LocalChannel)) or (self.move_files):
+        if self._should_move_files():
             logger.debug("Pushing start script")
             script_path = self.channel.push_file(script_path, self.channel.script_dir)
 
         logger.debug("Launching in remote mode")
-        # Bash would return until the streams are closed. So we redirect to a outs file
-        cmd = 'bash {0} > {0}.out 2>&1 & \n echo "PID:$!" '.format(script_path)
+        # We need to capture the exit code and the streams, so we put them in files. We also write
+        # '-' to the exit code file to isolate potential problems with writing to files in the
+        # script directory
+        #
+        # The basic flow is:
+        #   1. write "-" to the exit code file. If this fails, exit
+        #   2. Launch the following sequence in the background:
+        #      a. the command to run
+        #      b. write the exit code of the command from (a) to the exit code file
+        #   3. Write the PID of the background sequence on stdout. The PID is needed if we want to
+        #      cancel the task later.
+        #
+        # We need to do the >/dev/null 2>&1 so that bash closes stdout, otherwise
+        # channel.execute_wait hangs reading the process stdout until all the
+        # background commands complete.
+        cmd = 'echo - >{0}.ec && {{ {{ bash {0} 1>{0}.out 2>{0}.err ; ' \
```

Comment:
I worry that a non bash remote shell mess this up. Perhaps we should require bash in the channel.execute_wait calls.

Reply:
See discussion above with Ben.